lib/web/app/handler.go

```diff
@@ -140,29 +141,63 @@ func (h *Handler) handleForward(w http.ResponseWriter, r *http.Request, session
 	return nil
 }
 
+// handleForwardError when the forwarder has an error during the `ServeHTTP` it
+// will calll this function. This handler will then renew the session in order
+// to get "fresh" app servers, and then will forwad the request to the newly
+// created session.
+func (h *Handler) handleForwardError(w http.ResponseWriter, req *http.Request, err error) {
+	if !trace.IsConnectionProblem(err) && err != io.EOF {
+		w.WriteHeader(http.StatusInternalServerError)
+		w.Write([]byte(http.StatusText(http.StatusInternalServerError)))
+		return
+	}
+
+	session, err := h.renewSession(req)
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		w.Write([]byte(http.StatusText(http.StatusInternalServerError)))
+		return
+	}
+
+	session.fwd.ServeHTTP(w, req)
+}
+
 // authenticate will check if request carries a session cookie matching a
 // session in the backend.
 func (h *Handler) authenticate(ctx context.Context, r *http.Request) (*session, error) {
-	sessionID, err := h.extractSessionID(r)
+	ws, err := h.getWebSession(r)
 	if err != nil {
-		h.log.Warnf("Failed to extract session id: %v.", err)
+		h.log.Debugf("Failed to fetch application session: not found.")
 		return nil, trace.AccessDenied("invalid session")
 	}
 
-	// Check that the session exists in the backend cache. This allows the user
-	// to logout and invalidate their application session immediately. This
-	// lookup should also be fast because it's in the local cache.
-	ws, err := h.c.AccessPoint.GetAppSession(ctx, types.GetAppSessionRequest{
-		SessionID: sessionID,
-	})
+	// Fetch a cached session or create one if this is the first request this
+	// process has seen.
+	session, err := h.getSession(ctx, ws)
+	if err != nil {
+		h.log.Warnf("Failed to get session: %v.", err)
+		return nil, trace.AccessDenied("invalid session")
+	}
+
+	return session, nil
+}
+
+// renewSession based on the request, generates removes the session from cache
+// (if present), and generates a new one using the `getSession` flow (same as
+// in `authenticate`).
+func (h *Handler) renewSession(r *http.Request) (*session, error) {
+	ws, err := h.getWebSession(r)
 	if err != nil {
 		h.log.Debugf("Failed to fetch application session: not found.")
 		return nil, trace.AccessDenied("invalid session")
 	}
 
-	// Fetch a cached session or create one if this is the first request this
-	// process has seen.
-	session, err := h.getSession(ctx, ws)
+	// Remove the session from the cache, this will force a new session to be
+	// generated and cached.
+	h.cache.remove(ws.GetName())
```

Comment:
If multiple requests reach here simultaneously, what happens? Can one nuke the session created by the other?

Reply:
If they are from the same `AppSession`, both requests will remove the same session and generate a new one later.

> Can one nuke the session created by the other?
Yes. It will create a scenario that we already have in the authentication flow where two requests are executed at the same time when there is no cached session: one request will overwrite the session generated by the other.

Although subsequent requests are not affected by this scenario, do you think we should avoid it?

Reply:
Based on current code I think that handling this situation seems to be "safe". The  `h.cache.remove` called from other goroutine will remove previously  inserted by other goroutine but the new session will be inserted. 
The only concerns that comes to my mind is performance impact because newSession  flow tries to dial all app services to check it services are healthy. 

So maybe we should consider using something like https://pkg.go.dev/golang.org/x/sync@v0.0.0-20210220032951-036812b2e83c/singleflight that allows to block simultaneous code executed based on a key but this probably be done in a separate ticket/PR.